feat level2 구현

ZhenxiKim opened the pull request:
level 2의 기능들을 구현하였습니다.
- 금액 배팅
- 게임 시작
- 카드덱에서 카드를 뽑아 딜러/참가자에게 분배

---
[구현의도]

1. 딜러와 게임참가자에게 카드 분배 시 셔플된 카드를 카드 덱에서 FIFO로 뽑는 게 적절하다고 생각하였습니다. cardDeck 필드의 타입을 List -> Queue 자료구조로 변경하였습니다.
2. 클래스, 필드명에 자세하게 작명하여 불필요한 주석을 작성하지 않도록 작성하였습니다.
3. Card Deck 객체에 카드 분배 메소드를 구현하였으며, 분배할 카드 수를 매개변수로 설정하였습니다.
4. Card Deck에 카드를 초기화하는 기능은 initial block을 통해 구현하였습니다.
4.1. static block을 고민해 봤는데, static block은 main method 전에 초기화 된다고 하여 initial block을 선택했습니다. 다른 방식으로는 생성자에서 Card Deck을 초기화하는 메소드를 호출할 수 있습니다.

ZhenxiKim committed:
feat level2 구현

- 금액 배팅
- 게임시작
- 카드덱에서 카드를 뽑아 딜러/참가자에게 분배

f-lab-maverick commented on src/CardDeck.java:
```diff
@@ -1,21 +1,35 @@
+		cardDeck.addAll(shuffleCardList);
 	}
+
+	public List<Card> distributeCard(int count) {
+		List<Card> drawnCardList = new ArrayList<>();
+		for (int i = 0; i < count; i++) {
+			drawnCardList.add(cardDeck.poll());
+		}
+		return drawnCardList;
+	}
```
만약 count로 큰 숫자가 들어오거나, 0보다 작은 숫자가 들어오면 어떻게 될까요?

ZhenxiKim replied:
요구사항에 맞지않는 카드가 담긴 list가 리턴되거나 null이 리턴됩니다. 예외상황에 대한 처리를 추가해보겠습니다.

f-lab-maverick commented on src/Participant.java:
```diff
@@ -1,8 +1,23 @@
+		System.out.println("입력하신 금액 : " + battingMoney + moneyUnitVal);
+		super.setHoldingMoney(new Money(battingMoney, MoneyUnit.valueOf(moneyUnitVal)));
+
+	}
```
Participant 도메인에서 도메인과 관련 없는 Controller / View의 역할을 하는 메서드가 선언되었네요.
이렇게 비즈니스 로직이 아닌 입력을 받고, 출력을 하는 코드가 도메인 로직과 섞일 경우 추후에 이걸 변경하고자 할때 View와의 결합도로 인해 변경에 어려움이 생길 수 있습니다.

도메인은 순수하게 도메인 비즈니스 로직만 남기고, View 로직은 별도 코드로 분리해주세요.

ZhenxiKim replied:
Controller, Viewer클래스로 분리하여 도메인 클래스에는 도메인 관련 로직만 남기도록 수정하였습니다. 도메인 별로 view, controller 클래스가 생기다 보니 도메인 성격으로 분리할 수 있어 패키지 구조를 도메인 별로 묶었습니다.
-> MVC패턴과 같이  controller -> domain -> viewer 구조처럼 의존관계를 적용하는게 좋을지 고민이 됩니다. ( mvc 참고 : https://www.geeksforgeeks.org/mvc-design-pattern/)

f-lab-maverick replied:
MVC 패턴을 적용해보셔도 좋습니다~ 이런 작은 프로젝트에서도 패턴을 적용해보는 연습을 하시면 왜 MVC패턴이 나왔고 이게 널리 사용되는지도 체감해보실 수 있겠네요

f-lab-maverick commented on src/User.java:
```diff
@@ -3,9 +3,16 @@
+
+	public void setHoldingCardList(List<Card> holdingCardList) {
+		this.holdingCardList = holdingCardList;
+	}
```
setter를 사용하셨네요! setter 사용시 여러 문제가 발생될 수 있을텐데요, setter대신 동작 메서드로 변경해보시는건 어떠신가요?
아래 링크에 setter 대신 어떻게 코드를 작성할 수 있는지 잘 정리되어있습니다.
https://colabear754.tistory.com/173

ZhenxiKim replied:
값 변경의 목적이 드러나지 않는 setter 사용을 지양하고 변경 의도를 드러낼 수 있는 메서드로 변경하였습니다. 다만, 메소드 명을 짓는 일이 쉽지 않았습니다.. 🧛

f-lab-maverick replied:
이름짓는게 제일 어렵다고 하죠

f-lab-maverick reviewed (changes requested):
확인하였습니다. 피드백 확인 부탁드려요~

ZhenxiKim committed:
feat 피드백 반영

- controller, view로직 분리
- 매개변수 예외처리
- setter 사용을 지양하고 명확한 의도가진 메소드로 변경
- 도메인 별로 패키지 분리

f-lab-maverick commented on src/card/CardDeck.java:
```diff
@@ -25,6 +27,9 @@ public class CardDeck {
 	}
 
 	public List<Card> distributeCard(int count) {
+		if (count < 0 || count < 2) {
```
1. Magic Number라는 안티 패턴이 있습니다. [이 링크](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)를 확인해보세요.
2. 이렇게 조건절이 있을 경우 이 조건절이 무엇을 의미하는지 표현하기 어려울 수 있습니다. 당장에야 카드 수가 0~2장까지 요청이 와야한다는 간단한 요구사항이지만, 조건이 여러가지가 된다면 이게 어떤 의미인지 알기가 점점 어려워집니다. 그렇기 때문에 이런 조건절은 boolean을 리턴하는 private method화를 하여 메서드 이름으로 이게 어떤 조건을 의미하는지 표현해보시는것도 좋습니다.

f-lab-maverick commented on src/game/Game.java:
```diff
@@ -0,0 +1,40 @@
+		dealer.takeCard(cardDeck.distributeCard(2));
+	}
+
+	private Money getGameController() {
+		GameController gameController = new GameController();
+		return gameController.getParticipantBattingMoney();
```
메서드명과 리턴타입, 메서드 내용이 일치하지 않네요~

f-lab-maverick commented on src/game/Game.java:
```diff
@@ -0,0 +1,40 @@
+		return gameController.getParticipantBattingMoney();
+	}
+
+	private void getGameViewer(Money money) {
+		GameViewer gameViewer = new GameViewer();
+		gameViewer.printBattingMoney(money);
+	}
```
이것도 getter형식의 메서드이지만 메서드 내용과 메서드명이 일치하지 않습니다. 이런경우 휴먼 에러를 발생시키기 좋은 조건을 갖출 수 있습니다.

f-lab-maverick commented on src/user/User.java:
```diff
@@ -0,0 +1,26 @@
+ * @author jhkim
+ * @since 2023/06/18
+ */
+public class User {
+	private Money holdingMoney;
+	private List<Card> holdingCardList;
+
+	public void battingMoney(Money battingMoney) {
+		this.holdingMoney = battingMoney;
+	}
+
+	public void takeCard(List<Card> shuffledCardList) {
+		if (shuffledCardList.size() == 0) {
+			throw new IllegalArgumentException("딜러가 준 카드가 없습니다");
+		}
+		this.holdingCardList = shuffledCardList;
+	}
+}
```
메서드명을 setter가 아닌 행위메서드로 변경하셨네요~ 이 경우 내용도 행위메서드와 동일하게 setter와는 다른 행위를 하는것도 좋습니다. 마침 이유희님이 [이런 패턴](https://github.com/f-lab-edu/blackjack-game/blob/youhee/src/user/User.java)을 사용하고계신듯 하니 한번 확인해보는것도 좋을것 같네요~

User의 생성자가 없어서 holdingMoney, holdingCardList는 초기화해주기 전까지 null인 상태입니다.
명시적으로 `battingMoney`, `takeCard`를 호출해야 NPE에서 자유로워질 수 있겠네요.

이런 경우 setter보다는 생성자 패턴을 통해 초기화를 진행하는것이 좋습니다.

f-lab-maverick reviewed (changes requested):
약간 아쉬운 부분이 보여 approve는 하지 않았습니다.
여러 시도를 한 부분이 보이고 성장하는 것이 보입니다~ 피드백 반영 후 다시 리뷰 요청해주세요
